cluster/images/etcd-version-monitor/etcd-version-monitor.go

```diff
@@ -0,0 +1,219 @@
+	fs.StringVar(&etcdVersionScrapeURI, "etcd-version-scrape-uri", "http://localhost:2379/version", "URI to scrape etcd version info")
+	fs.StringVar(&etcdMetricsScrapeURI, "etcd-metrics-scrape-uri", "http://localhost:2379/metrics", "URI to scrape etcd metrics")
+	fs.DurationVar(&scrapeTimeout, "scrape-timeout", 15*time.Second, "Timeout for trying to get stats from etcd")
+}
+
+const (
+	namespace = "etcd"	// For prefixing prometheus metrics
+)
+
+// Initialize prometheus metrics to be exported.
+var (
+	etcdVersionFetchCount 	= prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name: "version_info_fetch_count",
+			Help: "Number of times etcd's version info was fetched, labeled by etcd's server binary and cluster version",
+		},
+		[]string{"server_version", "cluster_version"})
+	etcdGRPCRequestsTotal	= prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name: "grpc_requests_total",
+			Help: "Counter of received grpc requests, labeled by grpc method and grpc service names",
+		},
+		[]string{"grpc_method", "grpc_service"})
+)
+
+// Struct for unmarshalling the json response from etcd's /version endpoint.
+type EtcdVersion struct {
+	ServerVersion	string `json:"etcdserver"`
+	ClusterVersion	string `json:"etcdcluster"`
+}
+
+// Function for fetching etcd version info and feeding it to the prometheus metric.
+func getVersion() error {
+	// Create the get request for the etcd version endpoint.
+	req, err := http.NewRequest("GET", etcdVersionScrapeURI, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to create GET request for etcd version: %v", err)
+	}
+
+	// Send the get request and receive a response.
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	defer resp.Body.Close()
+	if err != nil {
+		return fmt.Errorf("Failed to receive GET response for etcd version: %v", err)
+	}
+
+	// Obtain EtcdVersion from the JSON response.
+	var version EtcdVersion
+	if err := json.NewDecoder(resp.Body).Decode(&version); err != nil {
+		return fmt.Errorf("Failed to decode etcd version JSON: %v", err)
+	}
+
+	// Record the info in the prometheus metric through labels.
+	etcdVersionFetchCount.With(prometheus.Labels{
+		"server_version": version.ServerVersion,
+		"cluster_version": version.ClusterVersion,
+	}).Inc()
+	return nil
+}
+
+// Periodically fetches etcd version info.
+func getVersionPeriodically(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			break
+		case <-time.After(scrapeTimeout):
+		}	
+		if err := getVersion(); err != nil {
+			glog.Errorf("Failed to fetch etcd version: %v", err)
+		}
+	}
+}
+
+// Struct for storing labels for gRPC request types.
+type GRPCRequestLabels struct {
+	GrpcMethod	string
+	GrpcService	string
+}
+
+// Function for fetching etcd grpc request counts and feeding it to the prometheus metric.
+func getGRPCRequestCount(lastRecordedCount *map[GRPCRequestLabels]float64) error {
```

Comment:
From an earlier discussion about this, it seemed better to have all version-related metrics (either the ones exposed directly by etcd or ones obtained through crunching (like the etcd version for example)) at one place. So we (re)expose some stuff by the version monitor. And yes, it would be great to have this upstream, but before that happens we need this for monitoring k8s cluster upgrades to 1.6.